Loader based API for image sources using ImageBitmap

ahocevar opened the pull request:
This pull request replaces #14841 and reverts #14444, which caused several problems.

My idea was to allow users to configure a plain `Image` source with a loader that takes extent, resolution and pixel ratio as arguments, and returns an image or a promise that resolves to an image. With such an API in place, users have a lot of flexibility for creating image sources with predefined loaders, their own loaders, or by chaining predefined loaders with custom loaders. In the simplest case, the usage looks like this:
```js
import {Image as ImageSource, createWMSLoader} from 'ol/source';

new ImageSource({
  loader: createWMSLoader({
    url: myWmsUrl,
    params: {LAYERS: '1'}
  });
});
```
The same works for static images:
```js
import {Image as ImageSource, createStaticLoader} from 'ol/source';

new ImageSource({
  loader: createStaticLoader({
    url: myStaticImageUrl,
    imageExtent: myStaticImageExtent
  })
});
```
As part of this effort, I wanted to provide a replacement solution for #14444. With a combination of the WMS loader and a custom image post-processing to return a canvas instead of the image, the result would have been the same as #14444. But in the process of doing so, I found out that the way we decode images in `ol/Image.js#listenImage()` leads to the problems described in #14436: using `Image.prototype.decode()` only guarantees the decoded image to be available for the next render frame. This is also the reason why the approach in #14841 required a weird combination of first rendering the image to a canvas and then storing the result as ImageBitmap, instead of rendering the image directly to an ImageBitmap: by the time the ImageBitmap would have been created, the decoded image was no longer available. It turned out that loading images with a promise that returns an ImageBitmap is as easy as
```js
image.src = './path/to/my/image.png';
return image.decode().then(() => createImageBitmap(image));
```

I created three different loaders in `ol/Image.js`. One that just loads the image (`load()`), one that loads and decodes it (`decode()`), and one that decodes the image and converts it to an `ImageBitmap` (`decodeToImageBitmap()`). The aforementioned loaders can be configured with a `load` function - it can be either of these three (only `load()` and `decodeToImageBitmap()` are marked as API for now), or a custom one.

This pull request is a major overhaul of the WMS (image and tile), ArcGISRest (image and tile), Static and MapGuide sources, with several changes to reuse more code between image and tile versions. The `Image` source is now useful by itself, thanks to the new `loader` option. The `ImageBase` and `Image` wrappers are now combined into one, with a simpler signature that supports the new loader functions.

The image layer renderer also gets an improvement - it now accepts different resolutions for x and y. This means that the code that used to resample an image in `ol/source/ImageStatic` could be removed, resulting in better performance and a lower memory footprint of static image sources that use images with different x- and y resolutions.

The new [wms-image-svg.html](https://deploy-preview-14945--ol-site.netlify.app/en/latest/examples/wms-image-svg.html) example shows how to take advantage of this.

The ArcGISRest loader can now also handle cases where the server returns an image with a different size as requested.

Fixes #14826
Fixes #14555
Fixes #14788
Fixes #14990

ahocevar commented on test/browser/spec/ol/source/TileWMS.test.js:
```diff
@@ -277,13 +277,11 @@ describe('ol/source/TileWMS', function () {
       expect(queryData.get('FORMAT')).to.be('image/png');
       expect(queryData.get('HEIGHT')).to.be('256');
-      expect(queryData.get('I')).to.be('517');
+      expect(queryData.get('I')).to.be('105');
```
This test was obiously wrong - with an image `WIDTH` of 256, `I` can never be greater than 256. If fixed the underlying issue, and this is also the reason for the different `BBOX` above, which was incorrect too.

ahocevar commented on src/ol/DataTile.js:
```diff
@@ -6,7 +6,7 @@ import TileState from './TileState.js';
 
 /**
- * @typedef {HTMLImageElement|HTMLCanvasElement|HTMLVideoElement} ImageLike
+ * @typedef {HTMLImageElement|HTMLCanvasElement|HTMLVideoElement|ImageBitmap} ImageLike
```
`ImageBitmap` works with the existing data tiles, and has the advantage of being decoded already - if created properly right after `HTMLImageElement#decode()`.

ahocevar commented on src/ol/Image.js:
```diff
@@ -22,159 +22,194 @@ import {listenOnce, unlistenByKey} from './events.js';
+ * @api
+ */
+
+/**
+ * Loader function used for image sources. Receives extent, resolution and pixel ratio as arguments.
+ * The function returns a promise for an  {@link import("./Image.js").ImageObject image object}.
+ *
+ * @typedef {function(import("./extent.js").Extent, number, number, (function(HTMLImageElement, string): void)=): import("./DataTile.js").ImageLike|ImageObject|Promise<import("./DataTile.js").ImageLike|ImageObject>} ImageObjectPromiseLoader
+ */
```
Returning an `ImageObject` object literal with the relevant image properties (image, extent, resolution and pixel ratio) rather than an `ImageWrapper` instance is convenient, and should also help avoid headache for 3rd party plugin developers, because they do not need an `ol` dependency when creating loaders.

ahocevar commented on test/browser/karma.config.cjs:
```diff
@@ -60,6 +60,9 @@ module.exports = function (karma) {
     ],
     proxies: {
       '/spec/': '/base/spec/',
+      '/wms': '/base/spec/ol/data/blank.png',
+      '/ImageServer/exportImage': '/base/spec/ol/data/blank.png',
+      '/MapServer/export': '/base/spec/ol/data/blank.png',
```
This configuration, in combination with the new `blank.png`, makes it easy to test the new loader with valid local urls and real image loading sequences.

ahocevar commented:
This is now ready for review. I know this is a bigger pull request than usual, and I'm sorry I wasn't able to split this into multiple commits. The modified code sections are deeply woven into image loading, source and renderer chains, so I ended up making just one big commit in addition to reverting the changes from #14444.

I apologize if this is unattractive to review. I tried to make it a bit easier by commenting some of the changes. Big thanks in advance for anyone who gives this a review, and of course I'm available for questions and discussions.

mike-000 commented on src/ol/source/static.js:
```diff
@@ -0,0 +1,48 @@
+ * @property {import("../extent.js").Extent} [imageExtent] Extent of the image in map coordinates.
+ * This is the [left, bottom, right, top] map coordinates of your image.
+ * @property {import("../size.js").Size} [imageSize] Size of the image in pixels. Usually the image size
+ * is auto-detected, so this only needs to be set for SVG images that have no implicit size.
```
I don't think this is useful.  An unsized SVG will still draw to canvas inconsistently and `createImageBitmap` will throw an error if it tries to use it.  Maybe time to deprecate this option.

An alternative to the `load` and `decodeToImageBitmap` methods for SVGs based on #14933 might be a better approach, but perhaps beyond the scope of the PR, something like

```
function svgLoad(image, src) {
  return new Promise((resolve, reject) => {
    let objectUrl;
    function handleLoad() {
      unlisten();
      resolve(image);
    }
    function handleError() {
      unlisten();
      reject(new Error('Image load error'));
    }
    function unlisten() {
      image.removeEventListener('load', handleLoad);
      image.removeEventListener('error', handleError);
      if (objectUrl) {
        URL.revokeObjectURL(objectUrl);
      }
    }
    image.addEventListener('load', handleLoad);
    image.addEventListener('error', handleError);

    if (
      image.crossOrigin !== undefined ||
      src.startsWith('data:image/svg+xml') ||
      location?.origin === new URL(src, location?.origin).origin
    ) {
      fetch(src)
        .then((response) => {
          if (response.headers.get('content-type').includes('image/svg+xml')) {
            return response.text().then((text) => {
              return new Blob([svgDimensions(text)], {type: 'image/svg+xml'});
            });
          }
          return response.blob();
        })
        .then((blob) => {
          objectUrl = URL.createObjectURL(blob);
          image.src = objectUrl;
        })
        .catch(() => {
          image.src = src;
        });
    } else {
      image.src = src;
    }
  });
}
```
(`src` would be required and could not be preset)

ahocevar replied:
Happy to remove the `imageSize` option. I only kept it for you, @mike-000.

ahocevar commented on test/browser/spec/ol/source/ImageStatic.test.js:
```diff
@@ -60,31 +69,49 @@ describe('ol/source/ImageStatic', function () {
     });
 
-    it('respects imageSize', function (done) {
+    it('Had an imageSize option which changed the imageExtent when used with incorrect image sizes', function (done) {
```
Keeping this test for historic reasons, to show that what `imageSize` did was neither useful nor what was advertised.

ahocevar commented:
Rebased to resolve conflicts.

mike-000 commented:
It is not clear how to use `ol/source/static#createLoader`.  Based on the WMS example I would expect to be able to replace an ImageStatic PNG with an SVG (e.g. the Image Reprojection example) simply by importing

```
import ImageSource from 'ol/source/Image.js';
import {createLoader} from 'ol/source/static.js';
import {load} from 'ol/Image.js';
```
and changing the source setup to
```
  const source = new ImageSource({
    loader: createLoader({
      url:
        'https://upload.wikimedia.org/wikipedia/commons/1/18/British_National_Grid.svg',
      crossOrigin: '',
      imageExtent: imageExtent,
      load: load,
    }),
    projection: 'EPSG:27700',
    interpolate: interpolate.checked,
  });
```
If it is any more complex than that another example would be useful.

ahocevar commented on src/ol/Image.js:
```diff
@@ -222,4 +257,67 @@ export function listenImage(image, loadHandler, errorHandler) {
+ * `HTMLImageElement` if `createImageBitmap()` is not supported.
+ * @api
+ */
+export function decodeToImageBitmap(image, src) {
```
Rename to `decode()`

ahocevar commented on src/ol/Image.js:
```diff
@@ -222,4 +257,67 @@ export function listenImage(image, loadHandler, errorHandler) {
+ * @param {string} [src] `src` attribute of the image. Optional, not required if already present.
+ * @return {Promise<HTMLImageElement>} Promise resolving to an `HTMLImageElement`.
+ */
+export function decode(image, src) {
```
Rename to `decodeFallback()`

ahocevar commented on test/browser/spec/ol/image.test.js:
```diff
@@ -1,85 +1,174 @@
-    );
-    image.addEventListener('change', function () {
-      if (image.getState() === ImageState.LOADED) {
+      setTimeout(function () {
```
Why was `setTimeout()` used here?

ahocevar replied:
This is just because the tests have moved and the diff is hard to read - this test already existed and had the same `setTimeout()`.

mike-000 commented on test/rendering/cases/reproj-image-stretched-interpolate-false/expected.png:
This is a regression of #12393.  Instead of preserving the X scale and scaling up the Y, pixels are being lost to scaling down, before the result is scaled up again.

It is clearer when using a bitmap from an SVG.
Stretched horizontally it looks as good as the source bitmap
![image](https://github.com/openlayers/openlayers/assets/49240900/8f0d9d76-015c-47e5-9b46-8301dc0e744e)

But instead of being stretched vertically the horizontal is compressed and pixels lost
![image](https://github.com/openlayers/openlayers/assets/49240900/8f7260b3-2724-41dd-8a32-c1ba78253890)

Instead of always passing the Y resolution to reprojection I think it should be the minimum?

ahocevar replied:
Changed to use `Math.min`.

tschaub commented on src/ol/Image.js:
```diff
@@ -222,4 +257,67 @@ export function listenImage(image, loadHandler, errorHandler) {
+export function load(image, src) {
+  if (src) {
+    image.src = src;
+  }
```
Maybe this doesn't make any difference given current implementations, but it feels like it would be safer to assign the `image.src` after registering the listeners below.

tschaub commented on src/ol/source/ImageArcGISRest.js:
```diff
@@ -151,81 +141,24 @@ class ImageArcGISRest extends ImageSource {
-      image.getPixelRatio() == pixelRatio &&
-      containsExtent(image.getExtent(), extent)
-    ) {
-      return image;
```
I wonder if this change will address #14990.

sunyuely replied:
I'm sorry, but this change does not address #14990.

sunyuely replied:
With the assistance of ahocevar, I have confirmed that the issue #14990 is caused by the limitations of the maxImageHeight and maxImageWidth properties in ArcGIS MapServer. Their default size is set to 4096.

ahocevar commented:
> It is not clear how to use `ol/source/static#createLoader`. Based on the WMS example I would expect to be able to replace an ImageStatic PNG with an SVG (e.g. the Image Reprojection example) simply by importing
> 
> ```
> import ImageSource from 'ol/source/Image.js';
> import {createLoader} from 'ol/source/static.js';
> import {load} from 'ol/Image.js';
> ```
> 
> and changing the source setup to
> 
> ```
>   const source = new ImageSource({
>     loader: createLoader({
>       url:
>         'https://upload.wikimedia.org/wikipedia/commons/1/18/British_National_Grid.svg',
>       crossOrigin: '',
>       imageExtent: imageExtent,
>       load: load,
>     }),
>     projection: 'EPSG:27700',
>     interpolate: interpolate.checked,
>   });
> ```
> 

This works now.

> The above code produces unexpected poor quality output for a scalable SVG when reprojected, and no output if the view projection is changed to EPSG:27700.

This has been fixed to return the same output as previously with an `ol/source/ImageStatic`.

> Taking a copy of the ImageStatic module to create a custom source with `decodeToImageBitmap` replaced by `load` produces good quality scaled output of an SVG without reprojection, but reprojected output is poor quality as it does not take advantage of the source image being scalable - this is due to it being drawn to the stitchContext - with a single source reprojection might be able to use the image directly instead of using stitchContext but that is probably beyond the scope here.

Definitely.

ahocevar commented:
I have addressed all comments, so this is now ready for a final review.

It's now also marked for the 8.0.0 milestone, and I have changed the upgrade notes to catecorize the removal of the `imageSize` option as backwards incompatible change.

tschaub reviewed (approved):
Really nice to see this come together.  Thanks for all the effort!

KyllianGautier commented:
Hello, I found a bug on this commit
I think the loader is not updated when the view changes (actually, when the projection changes)

https://github.com/openlayers/openlayers/issues/15161
